### ISSUE TITLE
Move from bounded wildcard to bounded generic

### ISSUE THREAD

**Issue by tir38** [bignerdranch/expandable-recycler-view#138]
`? extends ParentListItem` should become `T extends ParentListItem` to enforce type throughout classes `ExpandableRecyclerAdapter` and `ExpandableRecyclerAdapterHelper`

**Comment by paul-turner** [comment]
I'm curious on the why here.

What are the drawbacks of the bounded wildcard vs the bounded generic? It seems to me that the bounded wildcard allows for a little bit more flexibility in what the `ParentItemList` can be?

Willing to change it if it's deemed an improvement :)

**Comment by tir38** [comment]
if i understand the wildcard correctly, each use of `?` can be a different type. If' I'm using `ExpandableRecyclerAdapter` I can pass in one type in the constructor:

```
MyType extends ParentListItem
...

List<MyType> myParentItemList = new ...;
ExpandableRecyclerAdapter myAdapter = new ExpandableRecyclerAdapter(myParentItemList);
List<MyType> output = myAdapter.getParentItemList(); // this is not typesafe
```


But if we instead used a bounded generic. Then we can rely on the compiler to ensure the type saftey. We know the input type into the constructor is the same output type from 'getParentItemList()'

**Comment by jingibus** [comment]
I'm going to write a long complicated thing here which you don't need to read. It's fun if you want to, though. Next comment will be the summary recommendation.

1. It's true that each usage of `?` can be of a different `?`, but within `ParentItemList`, type safety is guaranteed. Check out the reference on wildcards here: https://docs.oracle.com/javase/tutorial/extra/generics/wildcards.html

    The example you provide here will not compile:

    <img width="516" alt="screen shot 2016-01-06 at 10 01 38 am" src="https://cloud.githubusercontent.com/assets/1508949/12145564/faa887fa-b45c-11e5-8146-491d00cd727b.png">

   Since types are a compile time construct, it *is* possible to use a combination of casts and wildcards to break type safety:

    <img width="549" alt="screen shot 2016-01-06 at 10 12 51 am" src="https://cloud.githubusercontent.com/assets/1508949/12145769/29ee3acc-b45e-11e5-9ade-6e212c6639b3.png">

    ...but obviously, that's a bad idea.

2. Having taken a look, it appears to me that there is no technical reason for the wildcard to be used here. Wildcards are there to make it possible interoperate between different generic types, not to make it possible to say, "Type X in this class is guaranteed to be subclass of ParentListItem." You can use wildcards to achieve that in some cases, but they're not required.

**Comment by jingibus** [comment]
Here's how it looks to me:

1. The nice thing about wildcards here is that they keep the user from needing to specify a bajillion type parameters. That's pretty nice.
2. The bad thing about wildcards here is that it breaks type safety on `getParentItemList`. Users of this class need to keep track of `parentItemList`'s type by hand; if they do that incorrectly, they're going to have a polluted collection.

So I see two sensible options:

1. Don't have `getParentItemList`. 
2. Don't use wildcards.

**Comment by paul-turner** [comment]
Hello, we've released 3.0.0-RC1 with a fix for this issue. 

I would love your feedback with any issues you have with the solution before we make 3.0.0 final. Check out the Pull Request that contains the fix for your issue (should be referenced above), and the [changelog](https://github.com/bignerdranch/expandable-recycler-view/releases/tag/v3.0.0-RC1) for a list of changes that have come with 3.0.0 (there are a lot of them!). 

If you have any issues feel free to re-open this issue with details of what you are seeing.
